include/osquery/http_client.h

```diff
@@ -0,0 +1,438 @@
+    boost::optional<std::string> client_private_key_file_;
+    boost::optional<std::string> ciphers_;
+    boost::optional<std::string> sni_hostname_;
+    boost::optional<std::string> proxy_hostname_;
+    boost::optional<std::string> remote_hostname_;
+    boost::optional<std::string> remote_port_;
+    long ssl_options_;
+    int timeout_;
+    bool always_verify_peer_;
+    bool follow_redirects_;
+    friend class Client;
+  };
+
+ public:
+  Client(Options const& opts = Options())
+      : client_options_(opts), r_(ios_), sock_(ios_), timer_(ios_) {}
+
+  /// HTTP put request method
+  Response put(Request& req,
+               std::string const& body,
+               std::string const& content_type = std::string());
+
+  /// HTTP post request method
+  Response post(Request& req,
+                std::string const& body,
+                std::string const& content_type = std::string());
+
+  /// HTTP get request method
+  Response get(Request& req);
+
+  /// HTTP head request method
+  Response head(Request& req);
+
+  /// HTTP delete_ request method
+  Response delete_(Request& req);
+
+  ~Client() {
+    closeSocket();
+  }
+
+ private:
+  /// Create Connection to server, if proxy option is set, connect to proxy
+  void createConnection();
+
+  /// Convert plain socket to ssl socket
+  void encryptConnection();
+
+  template <typename STREAM_TYPE>
+  void sendRequest(STREAM_TYPE& stream,
+                   Request& req,
+                   beast_http_response_parser& resp);
+
+  Response sendHTTPRequest(Request& req);
+
+  /// Handles HTTP request timeout
+  void timeoutHandler(boost_system::error_code const& ec);
+
+  /// Handles respone code, if requests completes or aborted due to timeout
+  void postResponseHandler(boost_system::error_code const& ec);
+
+  void closeSocket();
+
+ private:
+  Options client_options_;
+  boost_asio::io_service ios_;
+  boost_asio::ip::tcp::resolver r_;
+  boost_asio::ip::tcp::socket sock_;
+  boost_asio::deadline_timer timer_;
+  std::shared_ptr<ssl_stream> ssl_sock_;
+  boost_system::error_code ec_;
+  bool ssl_connection = false;
+  static const long SHORT_READ_ERROR = 0x140000dbL;
+  /// Setting the default port to squid proxy default port
+  static const int PROXY_DEFAUT_PORT = 3128;
+  static const int HTTP_DEFAULT_PORT = 80;
+  static const int HTTPS_DEFAULT_PORT = 443;
+};
+
+/**
+ * @brief HTTP request class
+ *
+ * This class is inherited from implemention(boost.beast) request class.
+ * It extends the functionality via providing uri parsing.
+ *
+ */
+template <typename T>
+class HTTP_Request : public T {
+ public:
+  /**
+   * @brief HTTP request header helper class
+   *
+   * constructor of this class takes (name, value)
+   * pair, which is used to set the request header
+   * of a http request with the help of overloaded
+   * function 'operator<<' of HTTP_Request class.
+   */
+  struct Header {
+    Header(const std::string& name, const std::string& value)
```

Comment:
Consider using these constructor signatures instead:
```
Header(beast_http::field name, boost::beast::string_param value);
Header(boost::beast::string_view name, boost::beast::string_param value);
```

Reply:
Actually I would just remove the `Header` class entirely, it doesn't seem to add any benefit that I can tell, and in most cases introduces unnecessary dynamic memory allocations.